### PR TITLE
Additional settlement logging

### PR DESCRIPTION
**Pull request by nlordell**
One small step towards #209

This PR adds some additional logging around received solutions. Specifically it will:
- Log **all** settlements from any solver before filtering
- Log settlements that are considered for the settlement competition (pre-simulation)
- Adds more context for why HTTP responses from solvers may be ignored

This is a kind of _"low effort"_ additional logging as it does not always specifically give reasons as to why specific solutions are being filtered (we just count if any solutions are being filtered per solver) and is a little brittle (boiler plate for each additional "filter" we add, doesn't solve a bit of "spaghetti" from filtering settlements at various point in the run loop). Still it is an improvement over what we have now since external solvers provide a single settlement - so we can tell for sure why an external solver settlement is being ignored.

### CODE REVIEW

File: crates/solver/src/solver/http_solver/settlement.rs [hunk 1]
```diff
@@ -27,16 +27,9 @@ pub async fn convert_settlement(
     context: SettlementContext,
     allowance_manager: Arc<dyn AllowanceManaging>,
 ) -> Result<Settlement> {
-    match IntermediateSettlement::new(settled.clone(), context, allowance_manager)
-        .await
-        .and_then(|intermediate| intermediate.into_settlement())
-    {
-        Ok(settlement) => Ok(settlement),
-        Err(err) => {
-            tracing::debug!("failed to process HTTP solver result: {:?}", settled);
```
Comment (nlordell) [inline crates/solver/src/solver/http_solver/settlement.rs:36]: hoisted this to the parent module to include context about the solver name.

File: crates/solver/src/driver/solver_settlements.rs [hunk 2]
```diff
@@ -147,6 +147,17 @@ pub fn retain_mature_settlements(
...
     let valid_settlement_indices = find_mature_settlements(min_order_age, &settlements);
 
+    for (_, (solver, settlement)) in settlements
+        .iter()
+        .enumerate()
+        .filter(|(i, _)| !valid_settlement_indices.contains(i))
+    {
+        tracing::debug!(
+            solver_name = %solver.name(), ?settlement,
+            "filtered settlement for not being mature",
```
Comment (MartinquaXD) [inline crates/solver/src/driver/solver_settlements.rs:157]: This one is super weird/complex. Not sure how helpful that message would be for external people.
Maybe we can log a link to the PR implementing this check? ;)
Not sure how to best communicate what this check is doing. 🤷‍♂️

Comment (nlordell) [reply crates/solver/src/driver/solver_settlements.rs:157]: I thought this was mostly for us to know if an external solver settlement gets filtered because all orders are "immature". We are being asked a lot "why is my settlement not being considered?". The idea with this PR is to mostly be able to answer that with a quick log query:
```
log: "auction{id=...}" and log: "solver_name=..."
```

One thing we can do with this message is make it more explicit as to what it means. I had used "for not being mature" to be as succinct as possible.

File: crates/solver/src/driver.rs [hunk 3]
```diff
@@ -513,18 +513,35 @@ impl Driver {
...
+                    }
+
                     // Do not continue with settlements that are empty or only liquidity orders.
+                    let settlement_count = settlement.len();
                     settlement.retain(solver_settlements::has_user_order);
+                    if settlement_count != settlement.len() {
+                        tracing::debug!(
+                            solver_name = %name,
+                            "settlement(s) filtered containing only liquidity orders",
+                        );
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/solver/src/driver.rs:538]: I'm not sure it helps in this case. The issue is for our single order solvers with X settlements, if we say N got filtered out for only having liquidity orders and M got filtered out for stable price violation, then we don't know (without further analysis) why a specific settlement was filtered (and including # of settlements filtered in each step doesn't help that much IMO). That being said, if you think this number is useful for debugging, adding it is super easy! LMK.

Since external order solvers only have one settlement, this log is informative enough for them.

File: crates/solver/src/driver.rs [hunk 4]
```diff
@@ -534,7 +551,14 @@ impl Driver {
                                 &self.token_list_restriction_for_price_checks,
                             )
                         });
+                        if settlement_count != settlement.len() {
+                            tracing::debug!(
+                                solver_name = %name,
+                                "settlement(s) filtered for violating stable price checks",
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/solver/src/driver.rs:557]: I was trying too hard to be succinct 😛. Will update.

File: crates/solver/src/driver/solver_settlements.rs (hunk 2 above)
Comment (MartinquaXD) [reply crates/solver/src/driver/solver_settlements.rs:157]: I thought the purpose for the improved logging was specifically for people testing with a local drive which would then have access to these logs. At least this is what it said in the slack [thread](https://cowservices.slack.com/archives/C0375NV72SC/p1652955145717999) which initiated this work.
I think the wording of this log is as good as it gets without having a wall of text there, though. :/

File: crates/solver/src/driver.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/solver/src/driver.rs:538]: > Since external order solvers only have one settlement, this log is informative enough for them.

Good point. 👍

**Review by MartinquaXD — APPROVED** [review]
I think there is still more we could do to give external people easy access to good information but this PR at least covers low hanging fruit so I guess it's fine.

File: crates/solver/src/driver/solver_settlements.rs (hunk 2 above)
Comment (nlordell) [reply crates/solver/src/driver/solver_settlements.rs:157]: Ah you are right. I misread the Slack post. I'll try to reword it to be more descriptive.

**Comment by nlordell** [comment]
> good information but this PR at least covers low hanging fruit so I guess it's fine.

I made sure not to close #209 as there are still improvements to make.
